gems/aws-sdk-s3/lib/aws-sdk-s3/directory_downloader.rb

```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+module Aws
+  module S3
+    # Raised when DirectoryDownloader fails to download objects from S3 bucket
+    class DirectoryDownloadError < StandardError
+      def initialize(message, errors = [])
+        @errors = errors
+        super(message)
+      end
+
+      # @return [Array<StandardError>] The list of errors encountered when downloading objects
+      attr_reader :errors
+    end
+
+    # @api private
+    class DirectoryDownloader
+      def initialize(options = {})
+        @client = options[:client] || Client.new
+        @executor = options[:executor] || DefaultExecutor.new
+        @options = options
+        @abort_download = false
+      end
+
+      attr_reader :client, :abort_download
+
+      # TODO: need to add progress tracker
+      def download(destination, bucket:, **options)
+        if File.exist?(destination)
+          raise ArgumentError 'invalid destination, expected a directory' unless File.directory?(destination)
+        else
+          FileUtils.mkdir_p(destination)
+        end
+
+        download_opts = options.dup
+        @bucket = bucket
+        @ignore_failure = download_opts.delete(:ignore_failure) || false
+        @errors = []
+
+        downloader = FileDownloader.new(client: client, executor: @executor)
+        producer = ObjectProducer.new(destination, build_producer_opts(download_opts))
+        producer.run
```

Comment:
It seems a little weird to need to call `run` on the producer here - see my comments below on the ObjectProducer's interface for a suggestion